### PR TITLE
Feature/microservices dashboard 1.0.1

### PR DESCRIPTION
**Pull request by andreasevers**

### CODE REVIEW

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+<img class="center-block" alt="Architecture" src="/img/microservices-dashboard/screenshot.png" width="70%">
+
+
+So you’ve jumped on the hype train, built a bunch of microservices, and got your first releases under your belt. 
+Now what?
+Our experiences taught us this is the easy part.
+With the newly obtained microservices freedom, teams easily plunge into a world of cowboys and unicorns.
+The big ball of mud is just around the corner.
+Panic, mayhem and chaos loom over the organisation, waiting for everything to spin out of control.
+Especially for any enterprise not residing in silicon valley, maintaining some sort of governance and compliancy is essential.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: Silicon Valley (capitalised)

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+What does a microservice architecture mean not just for the developer, but also for analysts and managers?
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: Would add an extra comma and _developers_ (plural):

> What does a microservice architecture mean**,** not just for the developer**s**, but also for analysts and managers?

Or maybe rephrase the sentence a bit since it looks a bit odd...
> What does a microservice architecture mean...? For the developers...? And for the analysts, managers, ...?

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+In theory a microservices architecture gives developers complete freedom to use whatever tools and frameworks they want inside their microservice.
+In practice, managers often want to slightly restrict that freedom to avoid complete chaos.
+It’s not uncommon for managers and architects to impose a set of choices developers can choose from, and goals the teams have to achieve.
+In order to facilitate recruitment and knowledge transfer, developers could be forced to choose between for instance Java or Javascript.
+Similarly, architects might enforce every microservice to have a quality gate in place and to have a technical debt less than five days. 
+
+Aside from the technical aspects inside a microservice, compliancy is even more crucial at the contract level.
+They should be defined according to an architectural vision and comply to standards across the organisation.
+Having the ability to track these compliancy regulations and quality assurances is a key enabler for management to push for technical excellence.
+Too many times managers are left clueless on how much effort is required to mature the architecture and which teams they have to chase.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: Too **often**...?

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+The dashboard on its own doesn’t really make a lot of sense when it’s not connected to the architecture it’s supposed to visualise.
+Aggregators pull in information which eventually gets translated into nodes and links on the dashboard.
+
+#### Health-indicators aggregator
+
+Spring Boot exposes production-ready endpoints through its Actuator module.
+The health endpoint returns information regarding the current health status of the application.
+The source of this information is a bunch of health indicators, describing various components and dependencies of the application.
+For instance, an application can have a dependency on a database, for which a health indicator will usually provide heath information to the health endpoint, indicating whether the database is up and the connection pool is not depleted.
```
Comment (tomvdbulck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: type: heath => health

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+Hence, health indicators describe an up-to-date relationship between the application they run on, and its dependencies.
```
Comment (tomvdbulck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: , and => remove the comma

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+While Spring Cloud ensures health indicators are automatically enabled when you are using service discovery, circuit breakers, a config server or other Spring Cloud services, health indicators don’t automatically describe a relationship between an application and another application it calls.
```
Comment (tomvdbulck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: , health indicators don’t automatically => start a new sentence here: . Health indicators ...

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+Managers want to know what is currently out there and what will be out there in the future.
+This doesn’t require a manual ticketing system, simply a smart dashboard with a timeline.
+
+### Analysts need to know what functionality is out there
+
+In order to reuse functionality and avoid duplication, functional analysts have a strong need for an overview of the current functional landscape.
+Knowing which resources are exposed by what microservices, and which events and messages are being sent back and forth between microservices and queues, can go a long way in helping analysts understand the state of the architecture. 
+
+Furthermore, impact analysis can significantly improve when an overview of components and how they are linked together is available to the analysts.
+Not only does it encourage analysts to identify and inform consumers of a changing service, it can help avoid introducing breaking changes due to negligence or ignorance.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: > ... it can help **to** avoid ...?

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+During troubleshooting, testers and analysts should be able to find out what services and backends are involved in a certain functional flow.
+
+Just like managers, functional analysts are interested in upcoming features and releases.
+On top of that, analysts can benefit from being able to define the future state of the ecosystem.
+Especially when multiple teams are working on similar functionality, it can be notoriously difficult to avoid duplication and breaches of bounded contexts.
+Using a dashboard to define what is coming up, can help to give an unambiguous view of the current and future landscape.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: > ... **it** can help to give **them** an unambiguous view

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+### Developers can benefit from a broader view as well
+
+In a devops organisation, developers have the responsibility to not only build but also run their services.
+Knowing which versions are deployed where, can assist developers in verifying their deployments are successful, but also to determine the versions of their dependencies.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: > ... in verifying **whether** their deployments are successful

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+In a microservices architecture, testing is absolutely crucial.
+As the primary benefit of microservices is faster time-to-market, changes happen all the time.
+Not only unit and integration testing is required, but also more advanced contract testing to act as a safety net.
+Consumer-driven-contract testing allows the consumer (the client) to define what he expects from the producer (the service), and ensure the producer validates that definition every time a change is made to the service.
+This allows the consumer to rest at ease, knowing the producer will remain backwards compatible or version accordingly, and gives the producer knowledge of who uses exactly what of its service.
+The latter gives the producer the chance to request consumers to update their service in case they are causing too much complexity on the producer’s side due to backwards compatibility.
+
+Tests like these document with guaranteed certainty relations between clients and services or services and services.
+Querying the contracts that define these relations offer a great source for the dashboard’s nodes and links between them.
```
Comment (tomvdbulck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: This sentence does not feel right: "offer a great source" (a source of what?)

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+Especially when multiple teams are working on similar functionality, it can be notoriously difficult to avoid duplication and breaches of bounded contexts.
+Using a dashboard to define what is coming up, can help to give an unambiguous view of the current and future landscape.
+
+### Developers can benefit from a broader view as well
+
+In a devops organisation, developers have the responsibility to not only build but also run their services.
+Knowing which versions are deployed where, can assist developers in verifying their deployments are successful, but also to determine the versions of their dependencies.
+
+A graphical dashboard can go a long way in providing clarity to developers.
+But most of all it can act as a hub for other tools and documentation available.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: Add a comma
> Most of all**,** ...

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md (hunk 1 above)
Comment (bart-blommaerts) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: +Our experiences taught us this is the easy part.

was the easy part?

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+This AngularJS application consumes endpoints exposed by its server component.
+It displays four columns: UI, Resources, Microservices and Backends.
+Each of these columns show nodes and links between them.
+The information for these links come from Spring Boot Actuator health endpoints, Pact consumer-driven-contract tests and hypermedia indexes, which are aggregated in the microservices-dashboard-server project.
+
+## The architecture
+
+<img class="center-block" alt="Architecture" src="/img/microservices-dashboard/architecture.png" width="50%">
+
+The dashboard currently consists out of an AngularJS 1.x application which speaks over HTTP to a Spring Boot application.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: > communicating

instead of 

> speaks over

?

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+The frontend uses D3.js to visualise the nodes in the four columns.
+We are currently in the process of completely rebuilding the frontend stack.
+Next version will be running on Angular 2, Typescript and EcmaScript 6.
+Most of D3.js will be taken care of by Angular 2 itself.
+
+Thanks to this refactor we’ll see the introduction of RxJS, making the frontend much more reactive in nature.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: This **refactoring**?

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+This aligns our frontend and backend components goals, since the backend is already running RxJava.
+Our efforts currently focus on replicating all functionality which is currently available in the dashboard, albeit with much more attention to quality and testing.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: > ... all functionalit**ies, currently available ...**

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+Subsequently new features and enhancements will be built on top of a much more mature and extendible frontend application.
+
+
+Our server component is powered by Spring Boot’s auto configuration.
+It’s a library which, once on a regular Spring Boot’s classpath, will automatically transform the Spring Boot application into a JSON graph-serving engine.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: > ... once on a regular Spring Boot’s classpath ...

Sounds odd, maybe rephrase to:

> ..., once on the classpath of a Spring Boot application, ...

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+It does so by using aforementioned RxJava.
+The idea of the server application is that it will fetch information from the microservices ecosystem, with for instance Spring Cloud's integration of service registries, and collect details of components and their relation within said ecosystem.
+Needless to say collecting this information requires a lot of outbound calls, and can pose a serious performance burden in case the landscape gets bigger.
+Making intelligent use of the system’s resources is absolutely necessary, and RxJava does just that.
+In the future we might migrate to Spring’s Reactor which has a more formal integration of the ReactiveX specification and better integration with Spring itself.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: I would make _Spring's Reactor_ clickable, directing the reader to more information about the new project.

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md (hunk 4 above)
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: Heath information? Typo?

> ... is up and whether the connection pool **hasn't been** depleted.

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+Hence, health indicators describe an up-to-date relationship between the application they run on, and its dependencies.
+While Spring Cloud ensures health indicators are automatically enabled when you are using service discovery, circuit breakers, a config server or other Spring Cloud services, health indicators don’t automatically describe a relationship between an application and another application it calls.
+Luckily Spring Boot has a very easy way of adding custom health indicators.
+As such, developers can add a health indicator the moment a remote service call is added to the application.
+
+
+> **What about real-time?**
+> 
+> Using health indicators we are certain the application calls another application programmatically.
+> This provides clarity in terms of what calls are in the code and therefore what dependencies exist across the applications.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: Would rephrase to:

> This provides clarity in terms of **the calls** in the code and therefore **the** dependencies **that** exist across the applications.

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+> However, using this method we aren’t sure whether this remote call is actually being executed at runtime.
+> These concerns are currently provided by other tools such as Twitter’s Zipkin.
+> In the future we will integration the dashboard with real-time traffic information from Zipkin or similar tooling.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: > In the future we will **integrate** ...

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+
+Once every microservice has an index resource, we can use service discovery to discover all the services, and fetch all the index resources to map out the landscape of resources.
+This is an excellent source for the dashboard, as it shows the relation between microservices and the RESTful resources they expose.
+
+#### Mappings aggregator
+
+Oftentimes, RESTful resources are exposed in a more traditional way (using REST level 2) without the added complexity of HATEOAS.
+While this is not fully REST compliant, it is most common among APIs using JSON over HTTP.
+Spring Boot offers a handy endpoint in their Actuator module, called the mappings endpoint.
+It describes all the resources which are exposed by the application when Spring MVC REST is used.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: > ... all the resources **exposed** by ....

File: _posts/2016-09-12-Microservices-Dashboard-1.0.1.md [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+While also describing Spring’s own resources, a simple filter allows us to deduct node and link information from these endpoints to visualise in the dashboard.
+
+#### Pact aggregator
+
+In a microservices architecture, testing is absolutely crucial.
+As the primary benefit of microservices is faster time-to-market, changes happen all the time.
+Not only unit and integration testing is required, but also more advanced contract testing to act as a safety net.
+Consumer-driven-contract testing allows the consumer (the client) to define what he expects from the producer (the service), and ensure the producer validates that definition every time a change is made to the service.
+This allows the consumer to rest at ease, knowing the producer will remain backwards compatible or version accordingly, and gives the producer knowledge of who uses exactly what of its service.
```
Comment (yannickdeturck) [inline _posts/2016-09-12-Microservices-Dashboard-1.0.1.md]: Maybe rephrase the last part a bit:
> ... producer knowledge of the services that are already in use.
